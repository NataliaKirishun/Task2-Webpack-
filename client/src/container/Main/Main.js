import React, { Component, Fragment } from 'react';
import styled from 'styled-components';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import EmptyResults from '../../components/EmptyResults/EmptyResults';
import Films from '../../components/Films/Films';
import Pagination from '../../common/Components/Pagination/Pagination';

class Main extends Component {

    getPagesArray(){
         const { albumClickHandler, movies, isFetched, total, currentPage, moviesPerPage } = this.props;
         const pagesCount = Math.ceil(total/moviesPerPage);
         let startPage;
         let lastPage;
         let pagesArray=[];
         let firstPageIsShown;
         let lastPageIsShown;
         if (currentPage<=5) {
            startPage = 1;
            lastPage = Math.min(Math.max(currentPage + 2, 5),pagesCount);
            firstPageIsShown = false;
            lastPageIsShown = (pagesCount>=5)? true : false;
         } else if (currentPage>=pagesCount-3) {
            startPage = currentPage-2;
            lastPage= pagesCount
            firstPageIsShown= true;
            lastPageIsShown= false;
         } else {
            startPage=currentPage-2;
            lastPage=currentPage+2;
            firstPageIsShown= true;
            lastPageIsShown= true;
         }
         for (let i=startPage; i<=lastPage; i++) {
            pagesArray.push(i);
         }
         return { pagesArray,pagesCount, firstPageIsShown, lastPageIsShown, moviesPerPage };
    }

  render() {
    const { albumClickHandler, movies, isFetched, total, currentPage, moviesPerPage } = this.props;

    const warningText = isFetched ? 'No films found for this request...' : 'Please, select your desired film.. ';
    return (
      <MainWrapper>
        {isFetched && movies.length
          ? (
          <Fragment>
            <Pagination
              paginationData={this.getPagesArray()}
              currentPage={currentPage}
              changePageHandler
              changePerPageHandler
            />
            <Films
              films={movies}
              albumClickHandler={albumClickHandler} />
            <Pagination
              paginationData={this.getPagesArray()}
              currentPage={currentPage}
              changePageHandler
              changePerPageHandler
            />
              </Fragment>
          )
          : <EmptyResults text={warningText} />
        }
      </MainWrapper>
    );
  }
}

export default connect(({movies, search})=>({
    movies: movies.movies,
    isFetched: movies.isFetched,
    total: movies.total,
    currentPage: search.currentPage,
    moviesPerPage: search.moviesPerPage,
}))(Main);

Main.propTypes = {
  searched: PropTypes.bool,
  filteredFilms: PropTypes.arrayOf(PropTypes.object),
  albumClickHandler: PropTypes.func,
};

const MainWrapper = styled.main`
    flex: 1 0 auto;

`;
