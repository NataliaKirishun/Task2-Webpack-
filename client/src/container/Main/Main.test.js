import React from 'react';
import { shallow , render } from 'enzyme';
import Main from './Main';

describe('Main Component Snapshot', () => {
    it('Main Component should render warning that no films were found when there are no such films', () => {
        const searched = true;
        const filteredFilms = [];
        const albumClickHandler= jest.fn();
        const component = shallow(<Main searched={searched} filteredFilms={filteredFilms} albumClickHandler={albumClickHandler} />);
        expect(component).toMatchSnapshot();
    });
    it('Main Component should render movies layout when there are several films found by the request', () => {
        const searched = true;
        const filteredFilms = [{
            "id":1891,
            "title":"The Empire Strikes Back",
            "tagline":"The Adventure Continues...",
            "vote_average":8.3,
            "vote_count":7414,
            "release_date":"1980-05-20",
            "poster_path":"https://image.tmdb.org/t/p/w500/6u1fYtxG5eqjhtCPDx04pJphQRW.jpg",
            "overview":"The epic saga continues as Luke Skywalker, in hopes of defeating the evil Galactic Empire, learns the ways of the Jedi from aging master Yoda. But Darth Vader is more determined than ever to capture Luke. Meanwhile, rebel leader Princess Leia, cocky Han Solo, Chewbacca, and droids C-3PO and R2-D2 are thrown into various stages of capture, betrayal and despair.",
            "budget":18000000,
            "revenue":538400000,
            "genres":["Adventure","Action","Science Fiction"],
            "runtime":124
        }];
        const albumClickHandler= jest.fn();
        const component = render(<Main searched={searched} filteredFilms={filteredFilms} albumClickHandler={albumClickHandler} />);
        expect(component).toMatchSnapshot();
    });

});

