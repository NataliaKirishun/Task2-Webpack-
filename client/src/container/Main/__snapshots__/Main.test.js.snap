// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Component Snapshot Main Component should render movies layout when there are several films found by the request 1`] = `
<main
  class="sc-jTzLTM etDbwe"
>
  <div
    class="sc-htpNat kfOUmV"
  >
    <div
      class="sc-bxivhb lhJBmx"
    >
      <div>
        <span
          id="filmsCount"
        >
          1
        </span>
        <span>
           movies found
        </span>
      </div>
      <form>
        <span>
          Sort by
        </span>
        <input
          class="sc-bdVaJa dDqVak"
          id="release"
          name="sort"
          type="radio"
          value="release date"
        />
        <label
          class="sc-bwzfXH loaEVr"
          for="release"
          id="active"
        >
          release date
        </label>
        <input
          class="sc-bdVaJa dDqVak"
          id="rating"
          name="sort"
          type="radio"
          value="rating"
        />
        <label
          class="sc-bwzfXH FnSOv"
          for="rating"
        >
          rating
        </label>
      </form>
    </div>
  </div>
  <div
    class="sc-gqjmRU deGoKe"
  >
    <div
      class="sc-VigVT hdSXxk"
    >
      <a
        class="sc-gzVnrw film-link jVAKsy"
      >
        <img
          class="sc-htoDjs hDxLSp"
          src="https://image.tmdb.org/t/p/w500/6u1fYtxG5eqjhtCPDx04pJphQRW.jpg"
        />
        <div
          class="sc-dnqmqq bEnhKB"
        >
          <h3
            class="sc-iwsKbI dcBvve"
          >
            The Empire Strikes Back
          </h3>
          <span>
            Adventure, Action, Science Fiction
          </span>
          <span
            class="sc-gZMcBi iJEmXo"
          >
            1980
          </span>
        </div>
      </a>
    </div>
  </div>
</main>
`;

exports[`Main Component Snapshot Main Component should render warning that no films were found when there are no such films 1`] = `
<styled.main>
  <Cockpit
    filmsCount={0}
    sortBy="release date"
    toggleSortBy={[Function]}
  />
  <EmptyResults
    text="No films found for this request..."
  />
</styled.main>
`;
